### ISSUE TITLE
Incorrect `reportOperatorIssue` error when comparing tuples

### ISSUE THREAD

**Issue by valorien** [microsoft/pyright#7733]
Tuple comparisons using the `<`/`<=`/`==`/`=>`/`>` operators are valid expressions in Python.
For some reason, Pylance regards these as errors and claims they are unsupported.

## Environment data

-   Language Server version: 2024.4.1
-   OS and version: Ubuntu 23.10 (WSL)
-   Python version: 3.12

## Code Snippet

```python
(7, 4) > (3, 2)   # Returns True
# Pylance: Operator ">" not supported for types "tuple[Literal[7], Literal[4]]" and "tuple[Literal[3], Literal[2]]"
```

**Comment by erictraut** [comment]
It looks like you've enabled type checking in pylance. The `reportOperatorIssue` is not enabled by default.

Pyright (the type checker upon which pylance is built) is working as intended here. The type of the LHS is inferred to be `tuple[Literal[7], Literal[4]]` and the type of the RHS is inferred to be `tuple[Literal[3], Literal[2]]`. For an explanation of how pyright infers types for literal expressions, refer to [this documentation](https://microsoft.github.io/pyright/#/type-inference?id=tuple-expressions).

The error comes from the evaluation of the `tuple.__gt__` call, which is provided by the typeshed stub `builtins.pyi`.

```python
class tuple(Sequence[_T_co]):
    def __gt__(self, value: tuple[_T_co, ...], /) -> bool: ...
```

The `tuple` class is a bit odd from a type-checking perspective because it's variadic. That means it supports an arbitrary number of type arguments, whereas the type definition makes it look like it accepts only one (`_T_co`). When applying typing rules, the single type parameter `_T_co` is effectively the union of the variadic type arguments. So, in your example, the type argument for LHS is effectively `Literal[7, 4]` and the type argument for the RHS is effectively `Literal[3, 2]`. These two types are not type compatible with each other, so pyright is correct (from a type analysis perspective) to generate an error here. I understand that this might be a dissatisfying answer because the expression does work fine at runtime.

Here are some options:
1. Use a `typing.cast` call to override the inferred type. `cast(tuple[int, int], (7, 4)) > (3, 2)`.
2. Use a `# type: ignore` or `# pyright: ignore[reportOperatorIssue]` comment to suppress the error.
3. Disable type checking if you find that it's too noisy.

**Comment by valorien** [comment]
Hi @erictraut, thank you for the prompt response.
I see your point about tuples being variadic, however this behavior is inconsistent with how `list` literals are treated:

```python
[7, 4] > [3, 2]  # No warning or error
``` 
Why should tuples issue a warning in this case when lists don't?

Thanks again.

**Comment by erictraut** [comment]
Tuples are immutable, so pyright retains literal types when inferring the type of a tuple expression. Lists are mutable, so literal types are not retained. The type of `(1, 2)` is inferred as `tuple[Literal[1], Literal[2]]` while the inferred type of `[1, 2]` is `list[int]`. For more details, refer to pyright's documentation [here](https://microsoft.github.io/pyright/#/type-inference?id=tuple-expressions).

**Comment by erictraut** [comment]
For an explanation of how pyright and mypy differ, refer to [this documentation](https://microsoft.github.io/pyright/#/mypy-comparison). There are many subtle choices design choices in both pyright and mypy that affect the interpretation of code. In both cases, you can find examples of where "valid operations" generate type errors. As a simple example, mypy generates an error for the following code even though it's valid and doesn't present any type safety issues.

```
x = 0
x = ""
```

The reason pyright infers tuples the way it does (which is admittedly different from mypy) is explained [here](https://microsoft.github.io/pyright/#/mypy-comparison?id=inference-of-tuple-expressions).

As with many design decisions in software, there are pros and cons to each of the design decisions in both pyright and mypy. 

I'll also note that there is an effort underway to standardize more behaviors between type checkers. The standard behaviors are spelled out by an official [Python typing standard](https://typing.readthedocs.io/en/latest/spec/index.html), which is being continually improved and augmented. Pyright more closely [conforms to the official spec](https://htmlpreview.github.io/?https://github.com/python/typing/blob/main/conformance/results/results.html) than mypy.

But yes, if mypy suits your needs better than pyright, it is available as an option.

**Comment by Andrej730** [comment]
Also came across this. 
I guess I previously noticed that it does work without errors when comparing tuples with first items matching, e.g. `tuple[x, y]` vs `tuple[x, y, z, w]`, which lead me to falsely expecting that `tuple[x, y]` vs `tuple[x, z]` would work too.

```python
print((4, 5) < (4, 5, 2, 0)) # True
# Operator ">" not supported for types
# "tuple[Literal[4], Literal[5]]" and "tuple[Literal[4], Literal[6], Literal[2], Literal[0]]"
print((4, 5) > (4, 6, 2, 0)) # False
```
